File: ai-service/utils/cache.py
# utils/cache.py
"""
Gerenciador de cache em memória.
"""

from datetime import datetime, timedelta
from typing import Any, Optional
import logging

logger = logging.getLogger(__name__)


class CacheManager:
    """
    Cache em memória com TTL (Time To Live).
    
    Exemplo:
        cache = CacheManager(ttl_seconds=1800)
        cache.set('key', {'data': 123})
        value = cache.get('key')  # Retorna {'data': 123}
    """
    
    def __init__(self, ttl_seconds: int = 3600):
        """
        Args:
            ttl_seconds: Tempo de vida do cache em segundos (padrão: 1h)
        """
        self.cache = {}
        self.ttl = timedelta(seconds=ttl_seconds)
        logger.info(f"✅ CacheManager iniciado (TTL: {ttl_seconds}s)")
    
    def get(self, key: str) -> Optional[Any]:
        """
        Busca valor do cache.
        
        Args:
            key: Chave de busca
        
        Returns:
            Valor armazenado ou None se expirado/inexistente
        """
        if key not in self.cache:
            logger.debug(f"🔍 Cache MISS: {key}")
            return None
        
        value, timestamp = self.cache[key]
        
        # Verifica expiração
        if datetime.now() - timestamp > self.ttl:
            del self.cache[key]
            logger.debug(f"⏰ Cache EXPIRED: {key}")
            return None
        
        logger.debug(f"✅ Cache HIT: {key}")
        return value
    
    def set(self, key: str, value: Any) -> None:
        """
        Armazena valor no cache.
        
        Args:
            key: Chave de armazenamento
            value: Valor a ser armazenado (qualquer tipo)
        """
        self.cache[key] = (value, datetime.now())
        logger.debug(f"💾 Cache SET: {key}")
    
    def delete(self, key: str) -> bool:
        """
        Remove valor do cache.
        
        Args:
            key: Chave a remover
        
        Returns:
            True se removido, False se não existia
        """
        if key in self.cache:
            del self.cache[key]
            logger.debug(f"🗑️ Cache DELETE: {key}")
            return True
        return False
    
    def clear(self) -> None:
        """Limpa todo o cache"""
        count = len(self.cache)
        self.cache.clear()
        logger.info(f"🧹 Cache limpo ({count} itens removidos)")
    
    def size(self) -> int:
        """Retorna número de itens no cache"""
        return len(self.cache)
    
    def cleanup_expired(self) -> int:
        """
        Remove itens expirados do cache.
        
        Returns:
            Número de itens removidos
        """
        now = datetime.now()
        expired_keys = [
            k for k, (_, ts) in self.cache.items()
            if now - ts > self.ttl
        ]
        
        for key in expired_keys:
            del self.cache[key]
        
        if expired_keys:
            logger.info(f"🧹 {len(expired_keys)} itens expirados removidos")
        
        return len(expired_keys)


# Instância global (singleton)
global_cache = CacheManager(ttl_seconds=1800)  # 30 minutos padrão

File: backend/utils/cache.js
// backend/utils/cache.js
/**
 * Cache em memória para Node.js (similar ao Python)
 * URGENTE: Performance crítica para aplicação web
 */

class CacheManager {
  constructor(ttlSeconds = 300) {
    // TTL padrão: 5 minutos (dados dinâmicos)
    this.cache = new Map();
    this.ttl = ttlSeconds * 1000; // Converte para ms
    console.log(`✅ CacheManager iniciado (TTL: ${ttlSeconds}s)`);
  }

  /**
   * Busca valor do cache
   */
  get(key) {
    const item = this.cache.get(key);
    
    if (!item) {
      return null; // Cache miss
    }

    const { value, timestamp } = item;
    const now = Date.now();

    // Verifica expiração
    if (now - timestamp > this.ttl) {
      this.cache.delete(key);
      return null; // Cache expired
    }

    return value; // Cache hit
  }

  /**
   * Armazena valor no cache
   */
  set(key, value, customTtl = null) {
    const ttl = customTtl ? customTtl * 1000 : this.ttl;
    this.cache.set(key, {
      value,
      timestamp: Date.now(),
      ttl
    });
  }

  /**
   * Remove valor do cache
   */
  delete(key) {
    return this.cache.delete(key);
  }

  /**
   * Limpa todo o cache
   */
  clear() {
    const size = this.cache.size;
    this.cache.clear();
    console.log(`🧹 Cache limpo (${size} itens removidos)`);
    return size;
  }

  /**
   * Remove itens expirados
   */
  cleanup() {
    const now = Date.now();
    let cleaned = 0;

    for (const [key, item] of this.cache.entries()) {
      const ttl = item.ttl || this.ttl;
      if (now - item.timestamp > ttl) {
        this.cache.delete(key);
        cleaned++;
      }
    }

    if (cleaned > 0) {
      console.log(`🧹 ${cleaned} itens expirados removidos do cache`);
    }

    return cleaned;
  }

  /**
   * Retorna tamanho do cache
   */
  size() {
    return this.cache.size;
  }

  /**
   * Invalida cache por padrão (ex: 'opportunities:*')
   */
  invalidatePattern(pattern) {
    let invalidated = 0;
    const regex = new RegExp(pattern.replace('*', '.*'));

    for (const key of this.cache.keys()) {
      if (regex.test(key)) {
        this.cache.delete(key);
        invalidated++;
      }
    }

    return invalidated;
  }
}

// Instância global (singleton)
const globalCache = new CacheManager(300); // 5 minutos padrão

// Limpeza automática a cada 1 minuto
setInterval(() => {
  globalCache.cleanup();
}, 60000);

module.exports = globalCache;



